CRM/ManualDirectDebit/Mail/DataCollector/Base.php

```diff
@@ -237,10 +281,18 @@ private function collectRecurringContributionRows() {
     $query = "
       SELECT 
         contribution.total_amount AS amount,
-        financial_type.name AS financial_type_name
+        financial_type.name AS financial_type_name,
```

Comment:
As explained before, you should add `civicrm_contribution.receive_date` to the select statement, so there's no need to calculate due dates manually.

Also, you could also sort by receive date, to make sure they're returned in the order you need.